apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,357 @@
+﻿// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Note: this will be in GAX, with the right copyright notice.
+
+using Google.Apis.Auth.OAuth2;
+using Grpc.Auth;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// Base class for API-specific builders.
+    /// </summary>
+    /// <typeparam name="TBuilder">The type of this builder, for chaining purposes</typeparam>
+    /// <typeparam name="TClient">The type of client created by this builder.</typeparam>
+    public abstract class ClientBuilder<TBuilder, TClient> where TBuilder : ClientBuilder<TBuilder, TClient>
+    {
+        /// <summary>
+        /// The endpoint to connect to, or null to use the default endpoint.
+        /// </summary>
+        public ServiceEndpoint Endpoint { get; private set; }
```

Comment:
Just to be sure, both ServiceEndpoint, ChannelCredentials and CallInvoker are inmutable, right? If they are not we should be consistent and follow the same pattern we are using for Settings on builder's implementations.

Reply:
ServiceEndpoint is. ChannelCredentials and CallInvoker are abstract classes - it's possible that the caller will provide a mutable version, but we don't have any way of cloning them, and I think that's basically a corner case a user could only get into deliberately.

Reply:
OK, if the user does that on their own, then, that's fine.